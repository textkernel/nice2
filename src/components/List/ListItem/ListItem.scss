$contexts: (neutral, brand, primary, accent, info, good, warning, bad);

.ListItem {
    transition: background-color var(--transition-duration);

    &__container {
        display: flex;
        align-items: baseline;
        padding: var(--spacing-4x);
    }

    &:hover:not(.ListItem--disabled) {
        background-color: var(--color-highlight-background);
    }

    &--isSelected, &--isHighlighted {
        background-color: var(--color-highlight-background);
    }

    &--clickable {
        cursor: pointer;
    }

    &--highlightContext {
        @each $context in $contexts {
            &_#{ $context }:hover:not(.ListItem--disabled) {
                background-color: var(--color-#{$context});
                color: var(--color-background);
            }

            &_#{ $context }.ListItem--isSelected {
                background-color: var(--color-#{$context});
                color: var(--color-background);
            }

            &_#{ $context }.ListItem--isHighlighted:not(.ListItem--disabled) {
                background-color: var(--color-#{$context});
                color: var(--color-highlight-background)
            }
        }
    }
}
