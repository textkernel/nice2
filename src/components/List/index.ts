export { List } from './List';
export { ListItem } from './ListItem';
export { ListActions } from './ListActions';
