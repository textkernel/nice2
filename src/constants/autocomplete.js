/* eslint-disable import/prefer-default-export */
export const NUMBER_OF_SUGGESTION_LOADING_PLACEHOLDERS = 5;
