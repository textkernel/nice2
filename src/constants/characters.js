// eslint-disable-next-line import/prefer-default-export
export const CROSS_CHAR = '\u00D7';
