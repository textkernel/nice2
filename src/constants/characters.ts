export const CROSS_CHAR = '\u00D7';
