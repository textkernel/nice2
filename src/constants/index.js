export const INPUT_TYPES = [
    'text',
    'password',
    'email',
    'date',
    'datetime-local',
    'month',
    'tel',
    'number',
    'url',
    'search',
    'time',
    'week',
];

export const LIST_NAVIGATION_DIRECTIONS = {
    UP: 'ArrowUp',
    DOWN: 'ArrowDown',
};

export const CROSS_CHAR = 'Ã—';

export * from './style-related';
export * from './autocomplete';
export * from './keyboard';
