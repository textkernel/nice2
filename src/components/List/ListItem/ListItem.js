import * as React from 'react';
import PropTypes from 'prop-types';
import { bem } from '../../../utils/bem';
import { Text } from '../../Text';
import { CONTEXTS } from '../../../constants';
import styles from './ListItem.scss';

const { block, elem } = bem('ListItem', styles);

export const ListItem = React.forwardRef((props, ref) => {
    const {
        children,
        isSelected,
        isHighlighted,
        onClick,
        disabled,
        highlightContext,
        ...rest
    } = props;
    const customBlockMod = { clickable: typeof onClick === 'function' };

    return (
        <li {...rest} ref={ref} {...block({ ...props, ...customBlockMod })}>
            <div onClick={onClick} role="presentation" {...elem('container', props)}>
                {React.Children.map(children, child =>
                    typeof child === 'string' ? <Text inline>{child}</Text> : child
                )}
            </div>
        </li>
    );
});

ListItem.displayName = 'ListItem';

ListItem.propTypes = {
    /** List Items */
    children: PropTypes.node,
    /** A function to be called if the item is clicked */
    onClick: PropTypes.func,
    /** Formats this item as selected  */
    isSelected: PropTypes.bool,
    /** Formats this item as selected  */
    isHighlighted: PropTypes.bool,
    /** Format this item as disabled */
    disabled: PropTypes.bool,
    /** formatting context when hovered or selected */
    highlightContext: PropTypes.oneOf([...CONTEXTS, 'default']),
};

ListItem.defaultProps = {
    children: null,
    onClick: null,
    isSelected: false,
    isHighlighted: false,
    disabled: false,
    highlightContext: 'default',
};
