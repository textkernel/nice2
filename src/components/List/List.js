import React, { useState, useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import bem from 'bem';
import ListItem from './ListItem';
import styles from './List.scss';
import { LIST_NAVIGATION_DIRECTIONS, ENTER_KEY } from '../../constants';

const { block, elem } = bem({
    name: 'List',
    classnames: styles,
    propsToMods: ['isDivided'],
});

const isListItem = element => element && element.type !== ListItem && element.type !== 'li';
const NAVIGATION_STEP_VALUES = {
    [LIST_NAVIGATION_DIRECTIONS.UP]: -1,
    [LIST_NAVIGATION_DIRECTIONS.DOWN]: 1,
};

const SCROLL_INTO_VIEW_SETTINGS = {
    block: 'nearest',
};

const List = React.forwardRef((props, ref) => {
    const [selectedIndex, setSelectedIndex] = useState(null);
    const [highlightedWithKeyboard, setHighlightedWithKeyboard] = useState(false);

    const highlightedListItem = useRef(null);

    const { children, isDivided, onNavigate, onSelect, isControlledNavigation, ...rest } = props;

    /**
     * Scroll list if it's necessary to make the highlighted item visible
     * every time selectedIndex was changed with the keyboard navigation
     */
    useEffect(() => {
        const hasScrollIntoViewFunction =
            highlightedListItem &&
            highlightedListItem.current &&
            highlightedListItem.current.scrollIntoView;

        if (highlightedWithKeyboard && hasScrollIntoViewFunction) {
            highlightedListItem.current.scrollIntoView(SCROLL_INTO_VIEW_SETTINGS);
            setHighlightedWithKeyboard(false);
        }
    }, [highlightedWithKeyboard, selectedIndex]);

    const getNextSelectedIndex = keyCode => {
        const stepValue = NAVIGATION_STEP_VALUES[keyCode];
        const nextSelectedIndex = selectedIndex + stepValue;

        // Return 0 index if nextSelectedIndex has negative value or selectedIndex hasn't been updated before
        if (nextSelectedIndex < 0 || selectedIndex === null) {
            return 0;
        }

        // Return last React.Children index if nextSelectedIndex is out of the right bound
        if (nextSelectedIndex >= children.length) {
            return children.length - 1;
        }

        // Return nextSelectedIndex without any changes for others cases
        return nextSelectedIndex;
    };

    const handleKeyDown = e => {
        // Update selectedIndex with arrow navigation and make onNavigate function callback
        if (e.key === LIST_NAVIGATION_DIRECTIONS.UP || e.key === LIST_NAVIGATION_DIRECTIONS.DOWN) {
            const nextSelectedIndex = getNextSelectedIndex(e.key);

            if (selectedIndex !== nextSelectedIndex) {
                e.preventDefault();
                setSelectedIndex(nextSelectedIndex);
                setHighlightedWithKeyboard(true);

                if (onNavigate) {
                    onNavigate(nextSelectedIndex, e.key);
                }
            }
        }

        // Imitate onClick event on Enter press and make onSelect function callback
        if (e.key === ENTER_KEY) {
            if (
                children[selectedIndex] &&
                children[selectedIndex].props &&
                children[selectedIndex].props.onClick
            ) {
                children[selectedIndex].props.onClick(e);

                if (onSelect) {
                    onSelect(selectedIndex);
                }
            }
        }
    };

    const handleMouseEnter = index => {
        if (selectedIndex !== index) {
            setSelectedIndex(index);
        }
    };

    return isControlledNavigation ? (
        <ul {...rest} ref={ref} {...block(props)}>
            {React.Children.map(children, child =>
                child ? React.cloneElement(child, elem('item', props)) : null
            )}
        </ul>
    ) : (
        // eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions, jsx-a11y/no-noninteractive-tabindex
        <ul {...rest} ref={ref} tabIndex="0" onKeyDown={handleKeyDown} {...block(props)}>
            {React.Children.map(children, (child, index) =>
                child
                    ? React.cloneElement(child, {
                          ...elem('item', props),
                          ref: index === selectedIndex ? highlightedListItem : null,
                          isHighlighted: index === selectedIndex,
                          onMouseEnter: () => handleMouseEnter(index),
                      })
                    : null
            )}
        </ul>
    );
});

List.displayName = 'List';

List.propTypes = {
    /** List items */
    children: (props, propName, componentName) => {
        const prop = props[propName];

        let error = null;
        React.Children.forEach(prop, child => {
            if (isListItem(child)) {
                error = new Error(
                    `'${componentName}' children should be of type 'ListItem' or 'li'.`
                );
            }
        });
        return error;
    },
    /** Adds dividing lines between the list items */
    isDivided: PropTypes.bool,
    /** onNavigate function callback. (selectedIndex: number, key: 'ArrowUp' || 'ArrowDown') */
    onNavigate: PropTypes.func,
    /** onSelect function callback. (selectedIndex: number) */
    onSelect: PropTypes.func,
    /** manage keyboard navigation externally */
    isControlledNavigation: PropTypes.bool,
};

List.defaultProps = {
    children: null,
    isDivided: false,
    onNavigate: null,
    onSelect: null,
    isControlledNavigation: false,
};

export default List;
