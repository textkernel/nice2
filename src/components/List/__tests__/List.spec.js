import React from 'react';
import toJson from 'enzyme-to-json';
import ListItem from '../ListItem';
import List from '../List';
import { LIST_NAVIGATION_DIRECTIONS } from '../../../constants';

describe('List component', () => {
    let consoleError;
    let wrapper;

    const itemNumbersArray = [0, 1, 2, 3, 4];
    const mockOnNavigate = jest.fn();
    const mockOnSelect = jest.fn();
    const mockOnClick = jest.fn();

    const navigateUp = () => wrapper.children().simulate('keyDown', { key: 'ArrowUp' });
    const navigateDown = () => wrapper.children().simulate('keyDown', { key: 'ArrowDown' });

    beforeEach(() => {
        consoleError = jest.spyOn(console, 'error').mockImplementationOnce(() => {});
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    describe('Initial rendering', () => {
        it('should render List correctly', () => {
            wrapper = mount(
                <List>
                    <ListItem>Item 1</ListItem>
                    <ListItem>Item 2</ListItem>
                </List>
            );

            expect(toJson(wrapper)).toMatchSnapshot();
            expect(wrapper.find('ul')).toHaveLength(1);
            expect(consoleError).not.toHaveBeenCalled();
        });
        it('should render correctly with ref', () => {
            const ref = React.createRef();
            mount(
                <List ref={ref}>
                    <ListItem>Item 1</ListItem>
                    <ListItem>Item 2</ListItem>
                </List>
            );

            expect(ref.current.props.Component.displayName).toEqual('List');
        });
        it('should warn if children are not ListItem nor li', () => {
            mount(
                <List>
                    <a href="/">Item 1</a>
                    <a href="/">Item 2</a>
                </List>
            );
            expect(consoleError).toHaveBeenCalled();
        });
    });

    describe('Keyboard navigation', () => {
        beforeEach(() => {
            wrapper = mount(
                <List onNavigate={mockOnNavigate} onSelect={mockOnSelect}>
                    {itemNumbersArray.map(number => (
                        <ListItem onClick={() => mockOnClick(number)}>Item ${number}</ListItem>
                    ))}
                </List>
            );
        });

        it('should not have any item highlighted from the start', () => {
            expect(mockOnNavigate).not.toHaveBeenCalled();
        });

        it('should update selectedIndex in both directions properly', () => {
            wrapper.children().simulate('click');
            navigateDown();

            expect(mockOnNavigate).toBeCalledWith(LIST_NAVIGATION_DIRECTIONS.DOWN);

            navigateDown();
            navigateUp();

            expect(mockOnNavigate).toBeCalledWith(LIST_NAVIGATION_DIRECTIONS.UP);
        });

        it('should not let highlighting selectedIndex got out of [0, itemsArray.length] bounds', () => {
            wrapper.children().simulate('click');
            navigateUp();
            navigateUp();

            expect(mockOnNavigate).toHaveBeenCalledTimes(1);

            itemNumbersArray.forEach(() => {
                navigateDown();
                navigateDown();
            });

            expect(mockOnNavigate).toHaveBeenCalledTimes(itemNumbersArray.length);
        });

        it('should call onSelect callback of the highlighted item', () => {
            wrapper.children().simulate('click');
            navigateDown();
            navigateDown();
            navigateDown();
            wrapper.children().simulate('keyDown', { key: 'Enter' });

            expect(mockOnClick).toBeCalledWith(2);
        });
    });
});
