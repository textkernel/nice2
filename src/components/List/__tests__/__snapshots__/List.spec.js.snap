// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component Initial rendering should render List correctly 1`] = `
<List
  isDivided={false}
  onNavigate={null}
  onSelect={null}
>
  <ul
    className="List"
    onKeyDown={[Function]}
    tabIndex="0"
  >
    <ListItem
      className="List__item"
      disabled={false}
      highlightContext="default"
      isHighlighted={false}
      isSelected={false}
      key=".0"
      onClick={null}
    >
      <li
        className="ListItem List__item"
      >
        <div
          className="ListItem__container"
          onClick={null}
          role="presentation"
        >
          <Text
            context="default"
            inline={true}
            key=".0"
            size="normal"
          >
            <span>
              Item 1
            </span>
          </Text>
        </div>
      </li>
    </ListItem>
    <ListItem
      className="List__item"
      disabled={false}
      highlightContext="default"
      isHighlighted={false}
      isSelected={false}
      key=".1"
      onClick={null}
    >
      <li
        className="ListItem List__item"
      >
        <div
          className="ListItem__container"
          onClick={null}
          role="presentation"
        >
          <Text
            context="default"
            inline={true}
            key=".0"
            size="normal"
          >
            <span>
              Item 2
            </span>
          </Text>
        </div>
      </li>
    </ListItem>
  </ul>
</List>
`;
